File: oss/clawpipe/python-sdk/clawpipe/telemetry.py
"""Telemetry -- per-request metrics tracking.

Tracks cost, tokens, latency, cache hits, and booster hits.
"""

from __future__ import annotations

import math
import time
from dataclasses import dataclass, field

from .types import TelemetrySnapshot


@dataclass
class RequestRecord:
    """A single recorded request."""

    provider: str
    model: str
    tokens_in: int
    tokens_out: int
    latency_ms: int
    cost_usd: float
    cached: bool
    boosted: bool
    timestamp: float = field(default_factory=time.time)


COST_TABLE: dict[str, float] = {
    "deepseek:deepseek-chat": 0.00014,
    "openai:gpt-4o-mini": 0.00015,
    "anthropic:claude-3-haiku": 0.00025,
    "openai:gpt-4o": 0.0025,
    "anthropic:claude-sonnet-4": 0.003,
    "anthropic:claude-opus-4": 0.015,
    "groq:llama-3.1-70b": 0.00059,
    "mistral:mistral-large": 0.002,
}


class Telemetry:
    """Per-request metrics tracking."""

    def __init__(self, max_records: int = 10_000) -> None:
        self._records: list[RequestRecord] = []
        self._max_records = max_records

    def estimate_cost(
        self, provider: str, model: str, tokens_in: int, tokens_out: int
    ) -> float:
        """Estimate cost in USD for a request."""
        key = f"{provider}:{model}"
        rate = COST_TABLE.get(key, 0.001)
        return ((tokens_in + tokens_out) / 1000) * rate

    def record(self, **kwargs: object) -> None:
        """Record a completed request."""
        if len(self._records) >= self._max_records:
            half = math.ceil(self._max_records * 0.5)
            self._records = self._records[-half:]
        self._records.append(RequestRecord(**kwargs))  # type: ignore[arg-type]

    def snapshot(self) -> TelemetrySnapshot:
        """Get aggregate telemetry snapshot."""
        r = self._records
        if not r:
            return TelemetrySnapshot()

        total_tokens_in = 0
        total_tokens_out = 0
        total_cost_usd = 0.0
        total_latency = 0
        cache_hits = 0
        booster_hits = 0
        model_stats: dict[str, dict[str, float]] = {}

        for rec in r:
            total_tokens_in += rec.tokens_in
            total_tokens_out += rec.tokens_out
            total_cost_usd += rec.cost_usd
            total_latency += rec.latency_ms
            if rec.cached:
                cache_hits += 1
            if rec.boosted:
                booster_hits += 1

            key = f"{rec.provider}:{rec.model}"
            if key not in model_stats:
                model_stats[key] = {"calls": 0, "cost": 0.0}
            model_stats[key]["calls"] += 1
            model_stats[key]["cost"] += rec.cost_usd

        top_models = sorted(
            [{"model": k, **v} for k, v in model_stats.items()],
            key=lambda x: x["calls"],
            reverse=True,
        )[:5]

        hit_rate = f"{(cache_hits / len(r) * 100):.1f}" if r else "0.0"

        return TelemetrySnapshot(
            total_requests=len(r),
            total_tokens_in=total_tokens_in,
            total_tokens_out=total_tokens_out,
            total_cost_usd=round(total_cost_usd, 4),
            total_saved_by_cache=cache_hits,
            total_saved_by_booster=booster_hits,
            avg_latency_ms=round(total_latency / len(r)),
            cache_hit_rate=f"{hit_rate}%",
            top_models=top_models,
        )

    def reset(self) -> None:
        """Reset all telemetry data."""
        self._records.clear()

File: oss/clawpipe/sdk/src/telemetry.ts
/**
 * Telemetry — per-request metrics tracking.
 *
 * Tracks cost, tokens, latency, cache hits, and booster hits.
 * Provides aggregate stats for reporting and debugging.
 */

import type { TelemetrySnapshot } from './types';

interface RequestRecord {
  provider: string;
  model: string;
  tokensIn: number;
  tokensOut: number;
  latencyMs: number;
  costUsd: number;
  cached: boolean;
  boosted: boolean;
  timestamp: number;
}

/** Cost per 1K tokens (input) for known models. */
const COST_TABLE: Record<string, number> = {
  'deepseek:deepseek-chat': 0.00014,
  'openai:gpt-4o-mini': 0.00015,
  'anthropic:claude-3-haiku': 0.00025,
  'openai:gpt-4o': 0.0025,
  'anthropic:claude-sonnet-4': 0.003,
  'anthropic:claude-opus-4': 0.015,
  'groq:llama-3.1-70b': 0.00059,
  'mistral:mistral-large': 0.002,
};

export class Telemetry {
  private records: RequestRecord[] = [];
  private maxRecords: number;

  constructor(maxRecords = 10_000) {
    this.maxRecords = maxRecords;
  }

  /** Estimate cost in USD for a request. */
  estimateCost(provider: string, model: string, tokensIn: number, tokensOut: number): number {
    const key = `${provider}:${model}`;
    // Conservative high fallback: overestimates rather than underestimates cost for unknown models.
    const rate = COST_TABLE[key] ?? 0.015;
    return ((tokensIn + tokensOut) / 1000) * rate;
  }

  /** Record a completed request. */
  record(entry: Omit<RequestRecord, 'timestamp'>): void {
    if (this.records.length >= this.maxRecords) {
      this.records = this.records.slice(-Math.ceil(this.maxRecords * 0.5));
    }
    this.records.push({ ...entry, timestamp: Date.now() });
  }

  /** Get aggregate telemetry snapshot. */
  snapshot(): TelemetrySnapshot {
    const r = this.records;
    if (r.length === 0) {
      return {
        totalRequests: 0, totalTokensIn: 0, totalTokensOut: 0,
        totalCostUsd: 0, totalSavedByCache: 0, totalSavedByBooster: 0,
        avgLatencyMs: 0, cacheHitRate: '0.0%', topModels: [],
      };
    }

    let totalTokensIn = 0, totalTokensOut = 0, totalCostUsd = 0;
    let totalLatency = 0, cacheHits = 0, boosterHits = 0;
    const modelStats = new Map<string, { calls: number; cost: number }>();

    for (const rec of r) {
      totalTokensIn += rec.tokensIn;
      totalTokensOut += rec.tokensOut;
      totalCostUsd += rec.costUsd;
      totalLatency += rec.latencyMs;
      if (rec.cached) cacheHits++;
      if (rec.boosted) boosterHits++;

      const key = `${rec.provider}:${rec.model}`;
      const existing = modelStats.get(key) ?? { calls: 0, cost: 0 };
      existing.calls++;
      existing.cost += rec.costUsd;
      modelStats.set(key, existing);
    }

    const topModels = Array.from(modelStats.entries())
      .map(([model, stats]) => ({ model, ...stats }))
      .sort((a, b) => b.calls - a.calls)
      .slice(0, 5);

    const hitRate = r.length > 0 ? ((cacheHits / r.length) * 100).toFixed(1) : '0.0';

    return {
      totalRequests: r.length,
      totalTokensIn,
      totalTokensOut,
      totalCostUsd: Math.round(totalCostUsd * 10000) / 10000,
      totalSavedByCache: cacheHits,
      totalSavedByBooster: boosterHits,
      avgLatencyMs: Math.round(totalLatency / r.length),
      cacheHitRate: `${hitRate}%`,
      topModels,
    };
  }

  /** Get records from the last N milliseconds. */
  recent(windowMs: number): RequestRecord[] {
    const cutoff = Date.now() - windowMs;
    return this.records.filter((r) => r.timestamp >= cutoff);
  }

  /** Total cost in the current window. */
  totalCostInWindow(windowMs: number): number {
    return this.recent(windowMs).reduce((sum, r) => sum + r.costUsd, 0);
  }

  /** Reset all telemetry data. */
  reset(): void {
    this.records = [];
  }
}
